provisional try to make minqlx independent from redis v2

mgaertne opened the pull request:
This is a first try to make minqlx redis less reliant on redis v2 being installed. I tried to incorporate backward and forward compatibility as much as possible. I did not yet try this out, though. Basic information point were the information from redis v3 page: https://pypi.org/project/redis/3.0.0/ I ignored SETEX, LREM, TTL, and PTTL changes. Not sure whether any plugin uses these features.

mgaertne committed:
provisional try to make minqlx independent from redis v2

This is a first try to make minqlx redis less reliant on redis v2 being installed. I tried to incorporate backward and forward compatibility as much as possible. I did not yet try this out, though. Basic information point were the information from redis v3 page: https://pypi.org/project/redis/3.0.0/ I ignored SETEX, LREM, TTL, and PTTL changes. Not sure whether any plugin uses these features.

em92 commented:
From discord:

> [15:48] ShiN0: the order of arguments in zincrby has been changed, instead of value, amount it's amount, value in redis >=3 (or the other way around). I checked for the second parameter filled in whether it's an int or float, and determined then whether it's a v2 call or a v3 call. Unfortunately if you call that function with a steam_id (it's an int) as value and 1 as amount, that logic might end up doing the wrong thing.

In this case, I suggest to make database class modifications, only if redis version is 3 or greater. For example:

```python
class Redis(AbstractDatabase):
   # ...
   if redis >= (3, 0, 0):
     def zincrby(self, name, value, amount=1):
       return self.r.zincrby(name, amount, value)

     #def mset(blablabla):
```

In that case Database's API will be like redis-py v2's API.

mgaertne commented:
I think it's more complicated than that. In my plugins, I incorporated some code that checks the redis version as well, and tries to call the functions in question in the right manner. That logic sort of collides with anything put into database.py.

em92 commented:
> In my plugins, I incorporated some code that checks the redis version as well, and tries to call the functions in question in the right manner.

Since you switched back to redis-py v2, maybe you will revert commits with redis version checks?

mgaertne commented:
That's a possibility. However, I grow concerned when the redis-py documentation and minqlx's adaptation go too far apart from each other. Ideally, one could check whether value and amount are both numeric, and if so, pick the smaller of the two to be the amount in the resulting call. I'm not sure whether that heuristic would be good-enough.

mgaertne committed:
Update database.py

added lrem and setex backwards compatability, and reworked zincrby to default on redis v2 behavior. If you want v3++ behavior, make sure to pass in value as str for all these functions.

em92 commented on python/minqlx/database.py:
```diff
@@ -313,3 +313,80 @@ def close(self):
+            return self.r.zincrby(name, value, amount)  # pylint: disable=W1114
+        return self.r.zincrby(name, amount, value)  # pylint: disable=W1114
+
+    def setex(self, name, value_or_time, time_or_value):
+        if not isinstance(value_or_time, (int, timedelta)):
+            value = value_or_time
+            time = time_or_value
+        else:
+            value = time_or_value
+            time = value_or_time
+
+        if redis.VERSION < (3, 0):
+            return self.r.setex(name, time, value)  # pylint: disable=W1114
+        return self.r.setex(name, value, time)  # pylint: disable=W1114
```
https://github.com/redis/redis-py/blob/3.5.3/redis/client.py#L1814-L1822
https://github.com/redis/redis-py/blob/2.10.6/redis/client.py#L1184-L1192

No need to derive setex.

mgaertne committed:
Update database.py
